src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs

```diff
@@ -0,0 +1,357 @@
+                var headers = new Dictionary<string, TextLine>();
+
+                // For option based code style diagnostic, try to find the .editorconfig key-value pair for the
+                // option setting.
+                var (optionNameOpt, optionValueOpt) = FindEditorConfigOptionKeyValueForRule();
+
+                // Check if an entry to configure the rule severity already exists in the .editorconfig file.
+                // If it does, we update the existing entry with the new severity.
+                var configureExistingRuleText = CheckIfRuleExistsAndReplaceInFile(optionNameOpt, result, headers);
+                if (configureExistingRuleText != null)
+                {
+                    return solution.WithAnalyzerConfigDocumentText(editorConfigDocument.Id, configureExistingRuleText);
+                }
+
+                // We did not find any existing entry in the in the .editorconfig file to configure rule severity.
+                // So we add a new configuration entry to the .editorconfig file.
+                var addMissingRuleText = AddMissingRule(optionNameOpt, optionValueOpt, result, headers);
+                if (addMissingRuleText != null)
+                {
+                    return solution.WithAnalyzerConfigDocumentText(editorConfigDocument.Id, addMissingRuleText);
+                }
+
+                // Bail out and return the old solution for error cases.
+                return solution;
+            }
+
+            private AnalyzerConfigDocument FindOrGenerateEditorConfig(ref Solution solution)
+            {
+                if (_project.AnalyzerConfigDocuments.Any())
+                {
+                    var diagnosticFilePath = PathUtilities.GetDirectoryName(_diagnostic.Location.SourceTree?.FilePath ?? _project.FilePath);
+                    if (!PathUtilities.IsAbsolute(diagnosticFilePath))
+                    {
+                        return null;
+                    }
+
+                    // Currently, we use a simple heuristic to find existing .editorconfig file.
+                    // We start from the directory of the source file where the diagnostic was reported and walk up
+                    // the directory tree to find an .editorconfig file.
+                    // In future, we might change this algorithm, or allow end users to customize it based on options.
+
+                    var bestPath = string.Empty;
+                    AnalyzerConfigDocument bestAnalyzerConfigDocument = null;
+                    foreach (var analyzerConfigDocument in _project.AnalyzerConfigDocuments)
+                    {
+                        var analyzerConfigDirectory = PathUtilities.GetDirectoryName(analyzerConfigDocument.FilePath);
+                        if (diagnosticFilePath.StartsWith(analyzerConfigDirectory) &&
+                            analyzerConfigDirectory.Length > bestPath.Length)
```

Comment:
i personally find this check confusing and can't convince myself it is right.  for example, say you have  `foo/a/b/c/d/e` and `foo/laskdjlksajdlakjsldkjsadlkjasldjasldjaskdjas`, the latter would preferable right?  but it's longer than the former.

Reply:
`diagnosticFilePath.StartsWith(analyzerConfigDirectory)` check ensures that we only consider ancestor directories of `diagnosticFilePath`. Only one of the above two paths in your example would be a valid candidated. Additionally, this is just a simple heuristic, and I fully anticipate we will have to fine tune it in future.